pv3_rust/src/pv.rs

```diff
@@ -0,0 +1,415 @@
+            packets[metadata_count as usize].size = nic.chunk_size;
+            packets[metadata_count as usize].buffer = unsafe { xsk_umem__get_data(nic.buffer, rx_desc_ptr.as_ref().unwrap().addr).cast::<u8>().sub(DEFAULT_HEADROOM as usize) };
+            packets[metadata_count as usize].private = unsafe { (rx_desc_ptr.as_ref().unwrap().addr - DEFAULT_HEADROOM as u64) as *mut c_void };
+            // println!("packet: {:?}", packets[metadata_count as usize]);
+            // packet_dump(&packets[metadata_count as usize]);
+            metadata_count += 1;
+        }
+
+        unsafe { xsk_ring_cons__release(&mut nic.rx, received); } // notify kernel of using filled slots in RX ring as much as **received
+
+        if metadata_count != received {
+            received = metadata_count;
+        }
+    }
+
+    unsafe {
+        if xsk_ring_prod__needs_wakeup(&mut nic.fq) != 0 {
+            libc::recvfrom(xsk_socket__fd(nic.xsk), std::ptr::null_mut::<c_void>(), 0, libc::MSG_DONTWAIT, std::ptr::null_mut::<libc::sockaddr>(), std::ptr::null_mut::<u32>()) ;
+        }
+    }
+
+    received
+}
+
+pub fn pv_send(nic: &mut PvNic, packets: &mut Vec<PvPacket>, batch_size: u32) -> u32 {
+    /* free packet metadata and UMEM chunks as much as the number of filled slots in cq. */
+    let mut cq_idx: u32 = 0;
+    let filled: u32 = unsafe{ xsk_ring_cons__peek(&mut nic.cq, batch_size, &mut cq_idx) }; // fetch the number of filled slots(the number of packets completely sent) in cq
+
+    if filled > 0 {
+        for _ in 0..filled {
+            unsafe { pv_free_(nic, *xsk_ring_cons__comp_addr(&nic.cq, cq_idx)); } // free UMEM chunks as much as the number of sent packets (same as **filled)
+            cq_idx += 1;
+        }
+        unsafe { xsk_ring_cons__release(&mut nic.cq, filled); } // notify kernel that cq has empty slots with **filled (Dequeue)
+    }
+
+     /* reserve TX ring as much as batch_size before sending packets. */
+    let mut tx_idx: u32 = 0;
+    let reserved: u32 = unsafe { xsk_ring_prod__reserve(&mut nic.tx, batch_size, &mut tx_idx) };
+    // println!("packets len: {}, tx reserved: {}, tx_idx: {}", packets.len(), reserved, tx_idx);
+
+    /* send packets if TX ring has been reserved with **batch_size. if not, don't send packets and free them */
+    if reserved < batch_size {
+        /* in case that kernel lost interrupt signal in the previous sending packets procedure,
+        repeat to interrupt kernel to send packets which ketnel could have still held.
+        (this procedure is for clearing filled slots in cq, so that cq can be reserved as much as **batch_size in the next execution of pv_send(). */
+        unsafe{
+            if xsk_ring_prod__needs_wakeup(&mut nic.tx) != 0 {
+                libc::sendto(xsk_socket__fd(nic.xsk), std::ptr::null(), 0, libc::MSG_DONTWAIT, std::ptr::null(), 0);
+            }
+        }
+        return 0;
+    }
```

Comment:
Why not sending (reserved count) packets and return number of sent packet?
That is what actually did in C version of PV

Reply:
my reply in the previous conversation is related to this conversation too

Reply:
But, Why not sending as much as possible.
Even if reserved amount is less than batch count, it still be able to send some packets and return that count

Reply:
Rust version has worked well like C version of PV3.0 since [e55b923](https://github.com/tsnlab/packetvisor/pull/44/commits/e55b923c201c1b54fb6458dc425d1e8d7b9bc334)

The application will be optimized as you pointed the issue when it needs to have higher performance